src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -367,135 +376,383 @@ public class SqlMutationEngine : IMutationEngine
+                                        parameters: parameters,
+                                        sqlMetadataProvider: sqlMetadataProvider);
+
+                            if (mutationResultRow is null || mutationResultRow.Columns.Count == 0)
+                            {
+                                if (context.OperationType is EntityActionOperation.Insert)
+                                {
+                                    if (mutationResultRow is null)
+                                    {
+                                        // Ideally this case should not happen, however may occur due to unexpected reasons,
+                                        // like the DbDataReader being null. We throw an exception
+                                        // which will be returned as an UnexpectedRestInsertOperationFailure
+                                        throw new DataApiBuilderException(
+                                            message: "An unexpected error occurred while trying to execute the query.",
+                                            statusCode: HttpStatusCode.InternalServerError,
+                                            subStatusCode: DataApiBuilderException.SubStatusCodes.UnexpectedRestInsertOperationFailure);
+                                    }
+
+                                    if (mutationResultRow.Columns.Count == 0)
+                                    {
+                                        throw new DataApiBuilderException(
+                                            message: "Could not insert row with given values.",
+                                            statusCode: HttpStatusCode.Forbidden,
+                                            subStatusCode: DataApiBuilderException.SubStatusCodes.DatabasePolicyFailure
+                                            );
+                                    }
+                                }
+                                else
+                                {
+                                    // This code block is reached when Update or UpdateIncremental operation does not successfully find the record to
+                                    // update. An exception is thrown which will be returned as a 404 NotFound response.
+                                    throw new DataApiBuilderException(message: "No Update could be performed, record not found",
+                                                                        statusCode: HttpStatusCode.NotFound,
+                                                                        subStatusCode: context.OperationType is EntityActionOperation.Update ? DataApiBuilderException.SubStatusCodes.UnexpectedRestUpdateOperationFailure
+                                                                                                                                             : DataApiBuilderException.SubStatusCodes.UnexpectedRestUpdateIncrementalOperationFailure);
```

Comment:
How can we expect/not expect a failure in this case? This should ideally be a user error. An error starting with `Unexpected` gives the sentiment that something went wrong with DAB. Can actually use the substatuscode `EntityNotFound`. This is being used in other places, one of them being here: https://github.com/Azure/data-api-builder/blob/ca5fa1463a38b3a0b5aebf31a4bff8aa64cebe73/src/Core/Resolvers/MsSqlQueryExecutor.cs#L273

Reply:
After more debugging, observed that `mutationResultRow.Columns` has a count of 0 when the operation is not able to find the record with the specified PK. Updated the status code to `EntityNotFound` for this case. 

Something to consider:  `EntityNotFound` is not completely accurate. Should a new status code like `ItemNotFound` be introduced to be accurate that the item with the given PK is not present. `EntityNotFound` gives an idea that the entity used in the request is incorrect.

`mutationResultRow` turns out to be `null` for unexpected errors - ideally should not happen but we need to account for this in the code however. So, using the `UnexpectedRestUpdateOperationFailure` and `UnexpectedRestUpdateIncrementalOperationFailure` status codes for this case.


Note: `mutationResultRow` being a `null` or `mutationResultRow.Columns.Count` being 0 are applicable only for the result set of `PerformMutationOperation`. They are inapplicable for the result set of `PerformUpsertOperation` - In this case, it would always contain a valid result set with data. Errors related to the result of `PerformUpsertOperation` being `null` or empty (column count == 0) are handled at the QueryExecutor/MsSqlQueryExecutor class.

Reply:
After investigation, I observed that we use `EntityNotFound` for 3 cases - a) Invalid Entity name (actual case where returning EntityNotFound is accurate), b) Item identified by PK is not present, c) Invalid PK field (the specified PK fields are not present in the entity definition) in the request.

Cases b) and c) needs additional changes. I can pick this up in a follow-up PR (In time for the second 0.9rc) . Don't want to hold this PR.

Reply:
Created [this issue](https://github.com/Azure/data-api-builder/issues/1809) for tracking.